Archive/Imo/Imo1982Q1.lean

```diff
@@ -0,0 +1,120 @@
+  rcases h with ( h₁ | h₂ )
+  · have : 2 * f 1 > 0 := by apply mul_pos (by norm_num) f1_pos
+    rw [← h₁] at this
+    exact lt_irrefl 0 this
+  · have : 2 * f 1 + 1 > 0 := by apply add_pos (mul_pos (by norm_num) f1_pos) (by norm_num)
+    rw [← h₂] at this
+    exact lt_irrefl 0 this
+
+lemma f₃ : f 3 = 1 := by
+    have h : f 3 = f 2 + f 1 ∨ f 3  = f 2 + f 1 + 1 := by apply hf.rel 2 1
+    rw [hf.f₁, hf.f₂, add_zero, zero_add] at h
+    have not_left : ¬ f 3 = 0 := by apply ne_of_gt hf.hf₃
+    rw [or_iff_right (not_left)] at h
+    exact h
+
+lemma superhomogeneity {m n : ℕ+} : ↑n * f m ≤ f (n * m) := by
+  induction n using PNat.recOn
+  case p1 => simp
+  case hp n' ih =>
+    have rel : f (n' * m + m) = f (n' * m) + f (m) ∨
+               f (n' * m + m) = f (n' * m) + f (m) + 1 := hf.rel (n' * m) m
+    calc
+    ↑(n' + 1) * f m = ↑n' * f m + f m := by rw [PNat.add_coe, add_mul, PNat.val_ofNat, one_mul];
+    _ ≤ f (n' * m) + f m := add_le_add_right ih (f m)
+    _ ≤ f (n' * m + m) := by
+      rcases rel with ( hl | hr)
+      · rw [hl]
+      · rw [hr]; nth_rewrite 1 [← add_zero (f m), ← add_assoc]; apply add_le_add_left (by norm_num)
+    _ = f ((n' + 1) * m) := by congr; rw [add_mul, one_mul]
+
+lemma superadditive {m n : ℕ+} : f m + f n ≤ f (m + n) := by
+  have h := hf.rel m n
+  rcases h with ( hl | hr )
+  · rw [hl]
+  · rw [hr]; nth_rewrite 1 [← add_zero (f n), ← add_assoc]; apply add_le_add_right (by norm_num)
+
+lemma superlinear {a b c d : ℕ+} : a * f b + c * f d ≤ f (a * b + c * d) := by
+  calc
+  a * f b + c * f d ≤ f (a * b) + f (c * d) := add_le_add hf.superhomogeneity hf.superhomogeneity
+  _ ≤ f (a * b + c * d) := by apply hf.superadditive
+
+lemma part_1 : 660 ≤ f (1980) := by
+  rw [← mul_one (660), ← PNat.val_ofNat 660, ← hf.f₃]; apply hf.superhomogeneity
+
+lemma part_2 : f 1980 ≤ 660 := by
+  have h : 5 * f 1980 + 33 * f 3 ≤ 5 * 660 + 33 := by
+    calc (5 : ℕ+) * f 1980 + (33 : ℕ+) * f 3 ≤  f (5 * 1980 + 33 * 3) := by apply hf.superlinear
+    _ = f 9999 := by rfl
+    _ = 5 * 660 + 33 := by rw [hf.f_9999]
+  rw [hf.f₃, mul_one] at *
+  -- from 5 * f 1980 + 33 ≤ 5 * 660 + 33 we show f 1980 ≤ 660
+  apply le_of_mul_le_mul_left (a := 5) (a0 := by norm_num)
+  apply le_of_add_le_add_right (a := 33)
+  exact h
+
+lemma main : f 1982 = 660 := by
+  have f_1980: f 1980 = 660 := le_antisymm hf.part_2 hf.part_1
+  have h : f 1982 = f 1980 + f 2 ∨ f 1982 = f 1980 + f 2 + 1 := hf.rel 1980 2
+  rw [f_1980, hf.f₂, add_zero] at h
+  rcases h with ( hl | hr)
+  · exact hl
+  · have h : 5 * f 1982 + 29 * f 3 + f 2 ≤ 3333 := by
+      calc
+      (5 : ℕ+) * f 1982 + (29 : ℕ+) * f 3 + f 2 ≤ f (5 * 1982 + 29 * 3) + f 2 := by
+        apply add_le_add_right hf.superlinear
+      _ ≤ f (5 * 1982 + 29 * 3 + 2) := by apply hf.superadditive
+      _ = f 9999 := rfl
+      _ = 3333 := by rw [hf.f_9999]
+    rw [hf.f₃, hf.f₂, hr, add_zero, mul_one] at h
+    -- 3334 ≤ 3333 a contradiction
+    simp at h
```

Comment:
Some very small tweaks.
```suggestion
  have f_1980 := hf.part_2.antisymm hf.part_1
  have h : f 1982 = f 1980 + f 2 ∨ f 1982 = f 1980 + f 2 + 1 := hf.rel 1980 2
  rw [f_1980, hf.f₂, add_zero] at h
  apply h.resolve_right
  intro hr
  have h : 5 * f 1982 + 29 * f 3 + f 2 ≤ 3333 := by
    calc
      (5 : ℕ+) * f 1982 + (29 : ℕ+) * f 3 + f 2 ≤ f (5 * 1982 + 29 * 3) + f 2 :=
        add_le_add_right hf.superlinear _
      _ ≤ f (5 * 1982 + 29 * 3 + 2) := hf.superadditive
      _ = 3333 := hf.f_9999
  rw [hf.f₃, hf.f₂, hr, add_zero, mul_one] at h
  -- 3334 ≤ 3333 a contradiction
  norm_num at h
```
You can keep the explicit typing in `f_1980` if you think it adds clarity.